File: Top_Interview_150/Array_String/cpp/134.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int canCompleteCircuit(vector<int> &gas, vector<int> &cost)
    {
        int start = 0;
        int station = 0;
        int s = 0;
        int n = gas.size();

        while (station - start < n)
        {
            s += gas[station] - cost[station];
            while (station - start < n && s < 0)
            {
                start--;
                s += gas[(start + n) % n] - cost[(start + n) % n];
            }
            station++;
        }

        if (s >= 0)
            return start >= 0 ? start : start + n;
        else
            return -1;
    }
};

File: Top_Interview_150/Array_String/python/134.py
from typing import List


class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:

        start = 0
        station = 0
        s = 0
        n = len(gas)

        while station - start < n:
            s += gas[station] - cost[station]
            while station - start < n and s < 0:
                start -= 1
                s += gas[start] - cost[start]
            station += 1

        if s >= 0:
            return start if start >= 0 else n + start
        return -1
